### PR TITLE
[PM-10599] Merge FeatureFlagConstants into FeatureFlags

### PR DESCRIPTION
**Pull request by KatherineInCode**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-10599

## 📔 Objective

This tries to create a unified, centralized feature flag system in the `FeatureFlag` and `ConfigService` objects. All feature flags and their corresponding string values (when not redundant) should be in the `FeatureFlag` enum now, and this also allows for feature flags to get an initial value that overrides the default (but not the remote server currently).

In terms of enabling things via compiler flags, that can be done by modifying the initial flags behind an `#if FEATURE` block.

I can see an argument for being able to use this mechanism to override the server as well, but I thought it best not to have the initial values do that (though if we want to change that, we can make them override values rather than default).

This means the cascade of feature flag settings is:

Debug Menu > Remote Configuration > Initial Value > Default Value > default Default Value (false/0/nil)

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

**Review by matt-livefront — APPROVED** [review]
Looks good! I had a few thoughts but nothing blocking.

File: BitwardenShared/Core/Platform/Services/ClientService.swift
Comment (KatherineInCode) [reply BitwardenShared/Core/Platform/Services/ClientService.swift:298]: My understanding is that there's been talk to do that in the SDK (and possibly to handle the entire `/config` endpoint logic and just surface up the feature flags for the apps), but it's lower priority than a lot of the other work, and we don't have anything concrete yet. I agree that ideally that would be surfaced by the SDK, though.

Comment (matt-livefront) [reply BitwardenShared/Core/Platform/Services/ClientService.swift:298]: Makes sense!

File: BitwardenShared/Core/Platform/Models/Enum/FeatureFlag.swift
Comment (fedemkr) [reply BitwardenShared/Core/Platform/Models/Enum/FeatureFlag.swift:57]: 🤔  I guess we can reduce the number of switch we need to configure all of this:
- `debugMenuFeatureFlags`: we can return directly here all the feature flags which its `raw` value doesn't start with `test-`.
- `name`: we can get the `raw` value and then change `-` for ` ` (space) and then apply `capitalized` so we get the same result as we do today without having to configure each flag on the switch. There's the special case of `enableCipherKeyEncryption` but in this case we can just check if there is no `-` then get the `raw` value and add a space each time we encounter a capital letter and then perform `capitalized` to uppercase the first letter as well.

Doing this we reduce these switches and is easier to maintain.

Comment (KatherineInCode) [reply BitwardenShared/Core/Platform/Models/Enum/FeatureFlag.swift:57]: I'm not sure I want to key off of `test` for determining if something goes in the debug menu, to give us the ability to control that a little more firmly in the future (in case there's something we don't want to put in there for some reason, such as cipher key encryption currently, since it's determined by completely different means than this system). But making that just an array of things in there is less cumbersome than a switch, I feel.

Doing something programmatic to get the name seems reasonable, since the only time the name should crop up is in the debug menu anyways.

I don't know that we can get rid of the `isRemotelyConfigured` per se, though we *could* convert it into an array of the flags that *are* remotely configured that can then be checked, which would reduce a switch.

I'm still fiddling around with the "struct to hold all this" idea to see how I like it (and discovering a bug in this code in the process 🙃 ), though. Definite tradeoffs, and I think ultimately it's not the better way to go. I'd rather move in the direction of having arrays of flags with certain properties, and we just check those—it does a better job of answering the question "what all flags have this property", which I think is a more useful question when looking at code than "what properties does this flag have", though I can see an argument either way. I'm still playing with it.

Comment (fedemkr) [reply BitwardenShared/Core/Platform/Models/Enum/FeatureFlag.swift:57]: IMO the debug menu stuff should be something not to spend much time to maintain, something simple with a bit of customization; additionally I believe that 90% of flags we'll introduce will be on that debug menu and not determined by some particular logic as cipher key encryption. 
Moreover, most feature flags are not supposed to live forever so I'd like to reduce the number of places we need to think of when we add/remove such flags.
Furthermore, about leaving the thing there as is / or as an array for further customization in the future in case we need it I'd just follow YAGNI and if in the future we encounter some weird flag with custom logic we need to add then we'll add it => I'd just key off `test-` and explicitly remove the cipher key encryption flag from the `debugMenuFeatureFlags`. I think with this we cover most cases therefore less effort to maintain as with array where we'd need to add/remove the key from the array every time a flag changes.

Comment (KatherineInCode) [reply BitwardenShared/Core/Platform/Models/Enum/FeatureFlag.swift:57]: Makes sense, and point taken on the debug menu. I can slim that down a fair bit for sure. I keep forgetting feature flags are meant to be transient 😂 

So 👍🏻 , there'll be some more commits coming soon for this.

Comment (fedemkr) [reply BitwardenShared/Core/Platform/Models/Enum/FeatureFlag.swift:53]: heheh great thanks Katherine!

File: BitwardenShared/Core/Platform/Models/Enum/FeatureFlag.swift
Comment (fedemkr) [inline BitwardenShared/Core/Platform/Models/Enum/FeatureFlag.swift:64]: Nice number choice hehe 😄 🎉

File: BitwardenShared/Core/Platform/Models/Enum/FeatureFlag.swift
Comment (fedemkr) [inline BitwardenShared/Core/Platform/Models/Enum/FeatureFlag.swift:96]: 🤔 I think this will show `EnableCipherKeyEncryption` instead of `Enable Cipher Key Encryption` but I guess it's fine for an internal debug menu not to add much more complexity to this. But if we'd like to add the case, I think for now it's just easier to add a specific if for that returning the value hardcoded than adding the logic to detect the uppercase letter to add the spacing.
```suggestion
        guard .enableCipherKeyEncryption != self else {
            return "Enable Cipher Key Encryption"
        }
        return rawValue.split(separator: "-").map(\.localizedCapitalized).joined(separator: " ")
```

**Review by fedemkr — APPROVED** [review]
Approving with just one small improvement but I'll leave it to you whether to include it or not as it's not strictly necessary.

File: BitwardenShared/Core/Platform/Models/Enum/FeatureFlag.swift
Comment (KatherineInCode) [reply BitwardenShared/Core/Platform/Models/Enum/FeatureFlag.swift:96]: Yeah, I thought about including something fancier, but since it's an internal debug menu *and* that particular flag is hidden from the menu already, I figured it wasn't worth the additional code (and attendant complexity). If we want to put it on the menu, we can add the hardcoded name when we do that.

**Pull request merged**
